### PR TITLE
Added rust-analyzer instructions for Helix

### PR DESCRIPTION
**Pull request by YohDeadfall**
That pull request adds information on how to configure Helix to use `rust-analyzer`, and moves the existing configuration to the `src/etc` directory as it's in the `rust` repository. Not adding instructions for other IDE because there's a link leading to the how-to for `rustc`.

### CODE REVIEW

File: src/etc/rust_analyzer_helix.toml
Comment (RalfJung) [inline src/etc/rust_analyzer_helix.toml:1]: I don't think we want this inside `src`.  Maybe just `etc`?

**Comment by YohDeadfall** [comment]
One more thing before finishing the pull request. `rust-analyzer` won't work just out of the box due to a custom toolchain created by the `./miri` script. Should then it be mentioned that `./miri toolchain` requires `-c rust-analyzer`? An alternative solution is to override the toolchain in the configuration files, or just add `-c rust-analyzer` directly to the script.

**Comment by RalfJung** [comment]
I'm using RA and I am not setting `-c rust-analyzer` anwhere. So not sure what you mean?

**Comment by YohDeadfall** [comment]
It depends how it was set up. For example, if I have no overrides in the config and no `rust-analyzer` for the `miri` toolchain then I can see that in the logs:

```
2024-10-03T23:29:00.949 helix_lsp::transport [ERROR] rust-analyzer err <- "error: 'rust-analyzer' is not installed for the custom toolchain 'miri'.\n"
2024-10-03T23:29:00.949 helix_lsp::transport [ERROR] rust-analyzer err <- "note: this is a custom toolchain, which cannot use `rustup component add`\n"
2024-10-03T23:29:00.949 helix_lsp::transport [ERROR] rust-analyzer err <- "help: if you built this toolchain from source, and used `rustup toolchain link`, then you may be able to build the component with `x.py`\n"
```

You don't have any issue, so I assume that you made a custom toolchain yourself with the analyzer and linked it, right?

**Comment by RalfJung** [comment]
> You don't have any issue, so I assume that you made a custom toolchain yourself with the analyzer and linked it, right?

No, I am just using `./miri toolchain` unchanged. With the vscode plugin, this just works. :shrug:

**Comment by YohDeadfall** [comment]
Interesting, maybe it's the plugin's magic, don't know... Then it can merged as is for now, or wait when I see what happens with VS Code.

Squash changes?

**Comment by YohDeadfall** [comment]
I checked why you don't get that error, and the reason is that the VS Code extension isn't a simple bridge between the IDE and `rust-analyzer`. It does a bit of smart job while bootstrapping the server ([soucre](https://github.com/rust-lang/rust-analyzer/blob/5982d9c420d0dc90739171829f0d2e9c80d98979/editors/code/src/bootstrap.ts#L58-L77)):

```ts
for (const workspaceFolder of vscode.workspace.workspaceFolders) {
    // otherwise check if there is a toolchain override for the current vscode workspace
    // and if the toolchain of this override has a rust-analyzer component
    // if so, use the rust-analyzer component
    const toolchainUri = vscode.Uri.joinPath(workspaceFolder.uri, "rust-toolchain.toml");
    if (await hasToolchainFileWithRaDeclared(toolchainUri)) {
        const res = spawnSync("rustup", ["which", "rust-analyzer"], {
            encoding: "utf8",
            env: { ...process.env },
            cwd: workspaceFolder.uri.fsPath,
        });
        if (!res.error && res.status === 0) {
            toolchainServerPath = earliestToolchainPath(
                toolchainServerPath,
                res.stdout.trim(),
                raVersionResolver,
            );
        }
    }
}
```

It means that `-c rust-analyzer` should be used for Helix users at least.

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -195,48 +195,35 @@ installed (`cargo install hyperfine`).
...
+
+Copy [`src/etc/rust_analyzer_settings.json`] to `.vscode/settings.json` in the project root directory.
+
+### Helix
+
+Copy [`src/etc/rust_analyzer_helix.toml`] to `.helix/languages.toml` in the project root directory.
+
+Since working on Miri requires a custom toolchain consider running `./miri toolchain` with the
+`-c rust-analyzer` flag, or set the `RUSTUP_TOOLCHAIN` environment variable according to
+[the documentation](https://rust-analyzer.github.io/manual.html#toolchain).
```
Comment (YohDeadfall) [inline CONTRIBUTING.md:211]: Added a note how to properly configure `rust-analyzer` for working on Miri. Another option would be following up the referenced documentation and setting the environment variable in the provided config, but I'm not sure about that. It seems too implicit, and in theory `rust-analyzer` might not be able to support everything.

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -195,48 +195,35 @@ installed (`cargo install hyperfine`).
...
-        "--message-format=json",
-    ],
-}
-```
+To configure `rust-analyzer` and the IDE for working on Miri, use one of the provided
+configuration files according to the instructions below.
+
+### Visual Studio Code
+
+Copy [`src/etc/rust_analyzer_settings.json`] to `.vscode/settings.json` in the project root directory.
```
Comment (RalfJung) [inline CONTRIBUTING.md:203]: The paths are still pointing to `src/`

**Comment by RalfJung** [comment]
> I checked why you don't get that error, and the reason is that the VS Code extension isn't a simple bridge between the IDE and rust-analyzer. It does a bit of smart job while bootstrapping the server ([soucre](https://github.com/rust-lang/rust-analyzer/blob/5982d9c420d0dc90739171829f0d2e9c80d98979/editors/code/src/bootstrap.ts#L58-L77)):

That logic will do nothing for Miri since we don't have  a `rust-toolchain.toml` file.

The actually relevant part is below:
```
// finally, use the bundled one
```
They simply bundle the server with the extension so it works without being installed via rustup.

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -195,48 +195,35 @@ installed (`cargo install hyperfine`).
...
+
+### Helix
+
+Copy [`src/etc/rust_analyzer_helix.toml`] to `.helix/languages.toml` in the project root directory.
+
+Since working on Miri requires a custom toolchain consider running `./miri toolchain` with the
+`-c rust-analyzer` flag, or set the `RUSTUP_TOOLCHAIN` environment variable according to
```
Comment (RalfJung) [inline CONTRIBUTING.md:210]:
```suggestion
Since working on Miri requires a custom toolchain, and Helix requires the language server
to be installed with the toolchain, you have to run `./miri toolchain -c rust-analyzer`
when installing the Miri toolchain. Alternatively, set the `RUSTUP_TOOLCHAIN` environment variable according to
```

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -195,48 +195,35 @@ installed (`cargo install hyperfine`).
...
+>
+> To keep the `rust-analyzer` configuration up-to-date, make a symbolic link to the provided file.
 
 > #### Note
 >
 > If you are [building Miri with a locally built rustc][], set
 > `rust-analyzer.rustcSource` to the relative path from your Miri clone to the
 > root `Cargo.toml` of the locally built rustc. For example, the path might look
-> like `../rust/Cargo.toml`.
+> like `../rust/Cargo.toml`. In addition to that replace `clippy` by `check` command
```
Comment (RalfJung) [inline CONTRIBUTING.md:222]:
```suggestion
> like `../rust/Cargo.toml`. In addition to that, replace `clippy` by `check`
```

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -195,48 +195,35 @@ installed (`cargo install hyperfine`).
...
+
+Since working on Miri requires a custom toolchain consider running `./miri toolchain` with the
+`-c rust-analyzer` flag, or set the `RUSTUP_TOOLCHAIN` environment variable according to
+[the documentation](https://rust-analyzer.github.io/manual.html#toolchain).
+
+> #### Hint
+>
+> To keep the `rust-analyzer` configuration up-to-date, make a symbolic link to the provided file.
 
 > #### Note
```
Comment (RalfJung) [inline CONTRIBUTING.md:217]: This now looks like the note is specific for Helix, which is not the case.

Any suggestion for how to avoid that confusion?

File: etc/rust_analyzer_settings.json [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+    "rust-analyzer.check.invocationLocation": "root",
+    "rust-analyzer.check.invocationStrategy": "once",
+    "rust-analyzer.check.overrideCommand": [
+        "env",
+        "MIRI_AUTO_OPS=no",
+        "./miri",
+        "clippy",
+        "--message-format=json",
+    ],
+    "rust-analyzer.cargo.buildScripts.invocationLocation": "root",
```
Comment (RalfJung) [inline etc/rust_analyzer_settings.json:17]:
```suggestion
    // Contrary to what the name suggests, this also affects proc macros.
    "rust-analyzer.cargo.buildScripts.invocationLocation": "root",
```

File: etc/rust_analyzer_helix.toml [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+]
+
+[language-server.rust-analyzer.config.check]
+invocationLocation = "root"
+invocationStrategy = "once"
+overrideCommand = [
+    "env",
+    "MIRI_AUTO_OPS=no",
+    "./miri",
+    "clippy", # make this `check` when working with a locally built rustc
```
Comment (RalfJung) [inline etc/rust_analyzer_helix.toml:18]: Why did you keep this comment here but remove it in the vscode file?

File: etc/rust_analyzer_settings.json
Comment (RalfJung) [inline etc/rust_analyzer_settings.json:1]: Shouldn't this be `rust_analyzer_vscode.json` to match the naming scheme?

File: etc/rust_analyzer_helix.toml [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+    "--message-format=json",
+]
+
+# Contrary to what the name suggests, this also affects proc macros.
+[language-server.rust-analyzer.config.buildScripts]
+enable = true
```
Comment (RalfJung) [inline etc/rust_analyzer_helix.toml:24]: Why is this needed here? The vscode file doesn't need it.

File: CONTRIBUTING.md (hunk 5 above)
Comment (YohDeadfall) [reply CONTRIBUTING.md:217]: I haven't decided how it is better yet. Moving it before the IDE paragraphs makes it too bold and the follow up seems disconnected. I will think about it.

File: etc/rust_analyzer_settings.json (hunk 6 above)
Comment (YohDeadfall) [reply etc/rust_analyzer_settings.json:17]: Comments aren't supported by JSON officially, so even when it was in a code block, these lines had a red background suggesting that it's invalid.

File: etc/rust_analyzer_settings.json
Comment (YohDeadfall) [reply etc/rust_analyzer_settings.json:1]: Yeah, makes sense. Originally I followed the naming from the main repo, but it definitely can be improved.

File: etc/rust_analyzer_settings.json (hunk 6 above)
Comment (RalfJung) [reply etc/rust_analyzer_settings.json:17]: It works in vscode though, so IMO that's a bug in github's JSON rendered and we should ignore it.

File: CONTRIBUTING.md [hunk 9]
```diff
@@ -195,48 +195,35 @@ installed (`cargo install hyperfine`).
...
 >
 > If you are [building Miri with a locally built rustc][], set
```
Comment (RalfJung) [inline CONTRIBUTING.md:219]: This link does not work, it seems. Probably pre-existing, but since you are touching this anyway...
```suggestion
> If you are building Miri with a locally built rustc, set
```

File: etc/rust_analyzer_helix.toml (hunk 8 above)
Comment (YohDeadfall) [reply etc/rust_analyzer_helix.toml:24]: Sorry, a leak from my config I made according to the documentation from `rustc`. The idea behind this setting is running build scripts for more precise code analysis. That's enabled for the `rust` repo.

I can mirror it for VS Code or remove it from here. What do you think?

Comment (RalfJung) [reply etc/rust_analyzer_helix.toml:24]: Aren't these enabled by default? AFAIK at least in vscode, they are.

File: etc/rust_analyzer_settings.json (hunk 6 above)
Comment (YohDeadfall) [reply etc/rust_analyzer_settings.json:17]: That's not a bug, but a feature (: Some parsers go against the specs and allow comments in JSON.

File: etc/rust_analyzer_helix.toml (hunk 8 above)
Comment (YohDeadfall) [reply etc/rust_analyzer_helix.toml:24]: According to the docs, yes, but in the settings from `rustc` it is enabled explicitly.

File: etc/rust_analyzer_settings.json (hunk 6 above)
Comment (RalfJung) [reply etc/rust_analyzer_settings.json:17]: What's relevant here is the vscode parser, and it (reasonably) supports comments. So we should use them -- it is quite useful to be able to explain configuration.

I'm not going to merge this with the comments removed.

File: etc/rust_analyzer_helix.toml (hunk 8 above)
Comment (RalfJung) [reply etc/rust_analyzer_helix.toml:24]: Not sure why they are doing that, but no need to copy it then -- let's just rely on the default.

File: etc/rust_analyzer_settings.json (hunk 6 above)
Comment (YohDeadfall) [reply etc/rust_analyzer_settings.json]: I already added them.

File: CONTRIBUTING.md (hunk 5 above)
Comment (YohDeadfall) [reply CONTRIBUTING.md]: Seems better now. Are you OK with it?

File: CONTRIBUTING.md [hunk 10]
```diff
@@ -195,48 +195,38 @@ installed (`cargo install hyperfine`).
...
+> If you are building Miri with a locally built rustc, set
 > `rust-analyzer.rustcSource` to the relative path from your Miri clone to the
 > root `Cargo.toml` of the locally built rustc. For example, the path might look
-> like `../rust/Cargo.toml`.
+> like `../rust/Cargo.toml`. In addition to that, replace `clippy` by `check`
+> in the `rust-analyzer.check.overrideCommand` setting.
+
+### Visual Studio Code
+
+Copy [`etc/rust_analyzer_settings.json`] to `.vscode/settings.json` in the project root directory.
```
Comment (RalfJung) [inline CONTRIBUTING.md:211]: What is the point of the brackets around the filename?

These do not become links like this.

File: CONTRIBUTING.md (hunk 5 above)
Comment (RalfJung) [reply CONTRIBUTING.md]: I would suggest to move the hint (about using a symlink) to the first section, the one about VSCode. People reading to-down will hopefully notice it then even if they are using a different IDE.

The note about "If you are building Miri with a locally built rustc" could move into the "Advanced configuration" section.

File: CONTRIBUTING.md (hunk 10 above)
Comment (YohDeadfall) [reply CONTRIBUTING.md:211]: I thought that it was a style in [rustc-dev-guide](https://rustc-dev-guide.rust-lang.org/building/suggested.html#helix), but it seems that there are just broken links.

File: CONTRIBUTING.md (hunk 5 above)
Comment (YohDeadfall) [reply CONTRIBUTING.md]: > People reading to-down will hopefully notice it then even if they are using a different IDE.

Maybe above the `### Visual Studio Code`? Then it wouldn't look like it applies only to VS Code, but also for Helix.

Comment (RalfJung) [reply CONTRIBUTING.md:225]: Then it is confusing because there haven't been any "files" yet that this comment refers to.

Comment (YohDeadfall) [reply CONTRIBUTING.md:225]: Make a symbolic link to the provided files below? Still the same place, but it point that the files are inside the IDE and Helix sections.

If you're not satisfied with that idea, I will just follow your suggestion to get it merged.

File: CONTRIBUTING.md [hunk 11]
```diff
@@ -195,48 +195,43 @@ installed (`cargo install hyperfine`).
...
 
 > #### Note
 >
-> If you are [building Miri with a locally built rustc][], set
+> If you are building Miri with a locally built rustc, set
 > `rust-analyzer.rustcSource` to the relative path from your Miri clone to the
 > root `Cargo.toml` of the locally built rustc. For example, the path might look
-> like `../rust/Cargo.toml`.
+> like `../rust/Cargo.toml`. In addition to that, replace `clippy` by `check`
+> in the `rust-analyzer.check.overrideCommand` setting.
```
Comment (RalfJung) [inline CONTRIBUTING.md:231]: It's odd to make this a "note" but the next thing here not. Please just make them two paragraphs.

File: CONTRIBUTING.md (hunk 5 above)
Comment (RalfJung) [reply CONTRIBUTING.md:225]: I don't understand the suggestion. But a forward reference will just be confusing I think, this hint can't come before the first concrete IDE configuration section.

File: CONTRIBUTING.md (hunk 11 above)
Comment (YohDeadfall) [reply CONTRIBUTING.md:231]: It was originally a note, I haven't touched that at all, just moved as it was asked (:

Comment (RalfJung) [reply CONTRIBUTING.md]: Yeah fair. :) But previously there was only a single IDE so we didn't need any of this structure. If you want I can push push some edits to your PR, that's probably a more effective way to do the final fine-tuning.

**Comment by RalfJung** [comment]
Okay, please squash the commits. :)

**Comment by RalfJung** [comment]
Thanks for the patience! I did some minor final tweaks.

@bors r+
